demos/full_waveform_inversion/full_waveform_inversion.py.rst

```diff
@@ -0,0 +1,292 @@
+
+    true_data_receivers = []
+    total_steps = int(final_time / dt) + 1
+    f = Cofunction(V.dual())  # Wave equation forcing term.
+    solver, u_np1, u_n, u_nm1 = wave_equation_solver(c_true, f, dt, V)
+    interpolate_receivers = Interpolator(u_np1, V_r).interpolate()
+
+    for step in range(total_steps):
+        f.assign(ricker_wavelet(step * dt, frequency_peak) * source_cofunction)
+        solver.solve()
+        u_nm1.assign(u_n)
+        u_n.assign(u_np1)
+        true_data_receivers.append(assemble(interpolate_receivers))
+
+Next, the FWI problem is executed with the following steps:
+
+1. Set the initial guess for the parameter ``c_guess``;
+
+2. Solve the wave equation with the initial guess velocity model (``c_guess``);
+
+3. Compute the functional :math:`J`;
+
+4. Compute the adjoint-based gradient of :math:`J` with respect to the control parameter ``c_guess``;
+
+5. Update the parameter ``c_guess`` using a gradient-based optimisation method, on this case the L-BFGS-B method;
+
+6. Repeat steps 2-5 until the optimisation stopping criterion is satisfied.
+
+**Step 1**: The initial guess is set as a constant field with a value of 1.5 km/s::
+
+    c_guess = Function(V).assign(1.5)
+
+
+.. image:: c_initial.png
+    :scale: 40 %
+    :alt: initial velocity model
+    :align: center
+
+
+To have the step 4, we need first to tape the forward problem. That is done by calling::
+
+    from firedrake.adjoint import *
+    continue_annotation()
+
+**Steps 2-3**: Solve the wave equation and compute the functional::
+
+    f = Cofunction(V.dual())  # Wave equation forcing term.
+    solver, u_np1, u_n, u_nm1 = wave_equation_solver(c_guess, f, dt, V)
+    interpolate_receivers = Interpolator(u_np1, V_r).interpolate()
+    J_val = 0.0
+    for step in range(total_steps):
+        f.assign(ricker_wavelet(step * dt, frequency_peak) * source_cofunction)
+        solver.solve()
+        u_nm1.assign(u_n)
+        u_n.assign(u_np1)
+        guess_receiver = assemble(interpolate_receivers)
+        misfit = guess_receiver - true_data_receivers[step]
+        J_val += 0.5 * assemble(inner(misfit, misfit) * dx)
+
+:class:`~.EnsembleReducedFunctional` is employed to recompute in parallel the functional and
+its gradient associated with the multiple sources (3 in this case)::
+
+    J_hat = EnsembleReducedFunctional(J_val, Control(c_guess), my_ensemble)
+
+The ``J_hat`` object is passed as an argument to the ``minimize`` function (see the Python code below).
+In the backend, ``J_hat`` executes simultaneously the computation of the cost function
+(or functional) and its gradient for each source based on the ``my_ensemble`` configuration. Subsequently,
+it returns the sum of these computations, which are input to the optimisation method.
+
+**Steps 4-6**: We can now to obtain the predicted velocity model using the L-BFGS-B method::
+
+    c_optimised = minimize(J_hat, method="L-BFGS-B", options={"disp": True, "maxiter": 1}, bounds=(1.5, 3.5))
+
```

Comment:
As I just discovered, minimize() will allgather the data for c_guess onto one spatial rank on each ensemble rank. We should probably at least mention here that this might cause problems for larger amounts of spatial parallelism.